addon/v1alpha1/types_clustermanagementaddon.go

```diff
@@ -175,10 +185,145 @@ type PlacementStrategy struct {
+	//
+	//   For example, the addon configs have spec hash A. The canary is passed and the lastKnownGoodConfigSpecHash
+	//   would be A, and all the selected clusters are rolling out to A.
+	//   Then the config spec hash changes to B. At this time, the clusters will continue rolling out to A.
+	//   When the rollout is done and canary passed B, the lastKnownGoodConfigSpecHash would be B and
+	//   all the clusters will start rolling out to B.
+	//
+	//   The canary placement does not have to be a subset of the install placement, and it is more like a
+	//   reference for finding and checking canary clusters before upgrading all. To trigger the rollout
+	//   on the canary clusters, you can define another rollout strategy with the type RollingUpdate, or even
+	//   manually upgrade the addons on those clusters.
+	//
+	// +kubebuilder:validation:Enum=UpdateAll;RollingUpdate;RollingUpdateWithCanary
+	// +kubebuilder:default:=UpdateAll
+	// +optional
+	Type string `json:"type"`
+
+	// Rolling update with placement config params. Present only if the type is RollingUpdate.
+	// +optional
+	RollingUpdate *RollingUpdate `json:"rollingUpdate,omitempty"`
+
+	// Rolling update with placement config params. Present only if the type is RollingUpdateWithCanary.
+	// +optional
+	RollingUpdateWithCanary *RollingUpdateWithCanary `json:"rollingUpdateWithCanary,omitempty"`
+}
+
+const (
+	// AddonRolloutStrategyUpdateAll is the addon rollout strategy representing apply the new configs to
+	// all the selected clusters at once.
+	AddonRolloutStrategyUpdateAll string = "UpdateAll"
+	// AddonRolloutStrategyRollingUpdate is the addon rollout strategy representing apply the new configs to
+	// all the selected clusters with the concurrency rate.
+	AddonRolloutStrategyRollingUpdate string = "RollingUpdate"
+	// AddonRolloutStrategyRollingUpdate is the addon rollout strategy representing wait and check
+	// if add-ons on the canary have applied the new configs, then apply the new configs to
+	// all the selected clusters with the concurrency rate.
+	AddonRolloutStrategyRollingUpdateWithCanary string = "RollingUpdateWithCanary"
+)
+
+// RollingUpdate represents the behavior to rolling update add-on configurations
+// on the selected clusters.
+type RollingUpdate struct {
+	// The maximum concurrently updating number of addons.
+	// Value can be an absolute number (ex: 5) or a percentage of desired addons (ex: 10%).
+	// Absolute number is calculated from percentage by rounding up.
+	// Defaults to 25%.
+	// Example: when this is set to 30%, once the addon configs change, the addon on 30% of the selected clusters
+	// will adopt the new configs. When the addons with new configs are healthy, the addon on the remaining clusters
+	// will be further updated.
+	// +kubebuilder:default:="25%"
+	// +optional
+	MaxConcurrency *intstr.IntOrString `json:"maxConcurrency,omitempty"`
```

Comment:
this has default value, do we need a pointer?

Reply:
add the pointer back, since when it's a structure, won't generate a default value.

Reply:
I think it should. it just cannot get default value if you create via client

Reply:
@qiujian16 Do we want to ignore the case when creating via client? and also remove the integration testing case?